File: hello-nodejs/src/client/util/PerformanceMonitor.js
/**
 * Performance monitoring for tracking upload/download metrics
 */

export class PerformanceMonitor {
  constructor(fileSize) {
    this.fileSize = fileSize;
    this.uploadStartTime = 0;
    this.uploadEndTime = 0;
    this.downloadStartTime = 0;
    this.downloadEndTime = 0;
  }

  startUpload() {
    this.uploadStartTime = performance.now();
  }

  endUpload() {
    this.uploadEndTime = performance.now();
  }

  startDownload() {
    this.downloadStartTime = performance.now();
  }

  endDownload() {
    this.downloadEndTime = performance.now();
  }

  getReport() {
    const uploadDurationMs = Math.round(
      this.uploadEndTime - this.uploadStartTime,
    );
    const downloadDurationMs = Math.round(
      this.downloadEndTime - this.downloadStartTime,
    );
    const totalDurationMs = uploadDurationMs + downloadDurationMs;

    // Calculate throughput in Mbps
    const fileSizeMb = (this.fileSize * 8) / (1024 * 1024);
    const uploadThroughputMbps =
      uploadDurationMs > 0 ? (fileSizeMb / uploadDurationMs) * 1000 : 0;
    const downloadThroughputMbps =
      downloadDurationMs > 0 ? (fileSizeMb / downloadDurationMs) * 1000 : 0;
    const averageThroughputMbps =
      totalDurationMs > 0 ? ((fileSizeMb * 2) / totalDurationMs) * 1000 : 0;

    return {
      uploadDurationMs,
      uploadThroughputMbps: parseFloat(uploadThroughputMbps.toFixed(2)),
      downloadDurationMs,
      downloadThroughputMbps: parseFloat(downloadThroughputMbps.toFixed(2)),
      totalDurationMs,
      averageThroughputMbps: parseFloat(averageThroughputMbps.toFixed(2)),
    };
  }
}

File: hello-python/src/client/util/performance_monitor.py
"""Performance monitoring for tracking upload/download metrics"""

import time
from typing import Dict


class PerformanceMonitor:
    """Performance monitor for tracking metrics"""
    
    def __init__(self, file_size: int):
        self.file_size = file_size
        self.upload_start_time = 0.0
        self.upload_end_time = 0.0
        self.download_start_time = 0.0
        self.download_end_time = 0.0
    
    def start_upload(self):
        """Start upload timer"""
        self.upload_start_time = time.perf_counter()
    
    def end_upload(self):
        """End upload timer"""
        self.upload_end_time = time.perf_counter()
    
    def start_download(self):
        """Start download timer"""
        self.download_start_time = time.perf_counter()
    
    def end_download(self):
        """End download timer"""
        self.download_end_time = time.perf_counter()
    
    def get_report(self) -> Dict[str, float]:
        """Generate performance report"""
        upload_duration_ms = round((self.upload_end_time - self.upload_start_time) * 1000)
        download_duration_ms = round((self.download_end_time - self.download_start_time) * 1000)
        total_duration_ms = upload_duration_ms + download_duration_ms
        
        # Calculate throughput in Mbps
        file_size_mb = (self.file_size * 8) / (1024 * 1024)  # Convert bytes to megabits
        upload_throughput_mbps = (file_size_mb / upload_duration_ms) * 1000 if upload_duration_ms > 0 else 0
        download_throughput_mbps = (file_size_mb / download_duration_ms) * 1000 if download_duration_ms > 0 else 0
        average_throughput_mbps = (file_size_mb * 2 / total_duration_ms) * 1000 if total_duration_ms > 0 else 0
        
        return {
            'upload_duration_ms': upload_duration_ms,
            'upload_throughput_mbps': round(upload_throughput_mbps, 2),
            'download_duration_ms': download_duration_ms,
            'download_throughput_mbps': round(download_throughput_mbps, 2),
            'total_duration_ms': total_duration_ms,
            'average_throughput_mbps': round(average_throughput_mbps, 2)
        }
